### DISCUSSION TITLE
Sorted keys

### DISCUSSION THREAD

**Discussion by koculu** [microsoft/FASTER#661] [Ideas]
I understand that Faster uses HashTable internally. That means the keys are not ordered.
Is there a desire to add a sorted data structure to Faster in the future?
Would it be good practice if I keep a sorted list of all keys in my application and continue to use Faster for lookups?
I would appreciate any other workaround suggestion.

**Comment by badrishc** [comment]
Yes, we would like to to add sorted secondary index support to FASTER, once the general secondary indexing framework (work in progress) is merged in. And yes, you can keep sorted list of keys and continue to use FASTER for lookups.

**Comment by koculu** [comment]
I have written an experimental B+ Tree backed by FasterKV with promising performance (at least it is faster than SortedSet at the moment.) I don't want to put more effort into it if native Range Queries are coming soon.

I want to learn more about the ongoing feature for secondary indexes and Range Queries.

Could you please share here relevant PRs, Discussions, and Documents?

If someone shares the deadline or estimation for these ongoing features, that would be pleasant.

Thank you.

**Reply by badrishc** [reply]
That is interesting. Our planned secondary index feature is slightly different from what you are mentioning. It is not a range index backed by FasterKV. Rather it is a capability of FasterKV as primary store, to forward all its updates to any indexing provider, so that that provider can index FASTER data references (e.g., in sorted order for range query support), as a secondary index that is directly queryable - in addition to the primary store. We do not have the PR or any discussion to share yet, once there is something we will add it to the main repo. There is no ETA as of now.

**Reply by cmadh** [reply]
@koculu is the source of your experimental B+ tree open source? My Team is pretty curious and would lovr to take a look.

**Comment by koculu** [comment]
@cmadh I have not open-sourced that because it was just experimental work. Anyway, I am attaching the code here for those interested. After my experiments and search on the topic, I decided to write an LSM tree with classical Lock-Based implementation. I am going to open-source that soon.
[Tenray.Graph.Console.zip](https://github.com/microsoft/FASTER/files/9002629/Tenray.Graph.Console.zip)

**Reply by cmadh** [reply]
Awesome, thank you very much @koculu ! 👍🏼👍🏼

**Comment by koculu** [comment]
@cmadh I have finally completed the LSM Tree for .NET and have made it open-source with an MIT license.
You may find it here: https://github.com/koculu/ZoneTree
